snapcraft/plugins/ruby.py

```diff
@@ -0,0 +1,142 @@
+        }
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return ['ruby-version', 'gems']
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._ruby_version = self.options.ruby_version
+        self._ruby_part_dir = os.path.join(self.partdir, 'ruby')
+        self._ruby_download_url = \
+            'https://cache.ruby-lang.org/pub/ruby/ruby-{}.tar.gz'.format(
+                self._ruby_version)
+        self._ruby_tar = Tar(self._ruby_download_url, self._ruby_part_dir)
+        self._gems = self.options.gems or []
+        self._install_bundler = False
+        self._ruby_version_dir = None
+
+        self.build_packages.extend(['gcc', 'g++', 'make', 'zlib1g-dev',
+                                    'libssl-dev', 'libreadline-dev'])
+
+        version_map = [('2.%d.[0-9]' % i, '2.%d.0' % i) for i in range(5)]
+
+        for version_regex, version_dir in version_map:
+            if re.compile(version_regex).match(self._ruby_version):
+                self._ruby_version_dir = version_dir
+                break
+        # Raise exception when version_dir not found
+        if not self._ruby_version_dir:
+            raise SnapcraftEnvironmentError(
+                'Ruby version dir not in version_map.')
+
+    def pull(self):
+        super().pull()
+        os.makedirs(self._ruby_part_dir, exist_ok=True)
+        self._ruby_tar.download()
+        self._ruby_install(builddir=self._ruby_part_dir)
+        self._gem_install()
+        if self._install_bundler:
+            self._bundle_install()
+
+    def env(self, root):
+        env = super().env(root)
+        env.append('PATH={}:{}'.format(
+            os.path.join(root, 'bin'), os.environ['PATH']))
+        env.append('RUBYPATH={}'.format(os.path.join(root, 'bin')))
+        rubydir = os.path.join(root, 'lib', 'ruby')
+        rubylib = os.path.join(rubydir, self._ruby_version_dir)
+        env.append('RUBYLIB={}:{}'.format(
+            rubylib, os.path.join(
+                rubylib, '{}-linux'.format(platform.machine()))))
+        env.append('GEM_HOME={}'.format(
+            os.path.join(rubydir, 'gems', self._ruby_version_dir)))
+        env.append('GEM_PATH={}'.format(
+            os.path.join(rubydir, 'gems', self._ruby_version_dir)))
+        return env
+
+    def _ruby_install(self, builddir):
+        self._ruby_tar.provision(
+            builddir, clean_target=False, keep_tarball=True)
+        self.run(['./configure', '--disable-install-rdoc', '--prefix=/'],
+                 cwd=builddir)
+        self.run(['make', '-j{}'.format(self.parallel_build_count)],
+                 cwd=builddir)
+        self.run(['make', 'install', 'DESTDIR={}'.format(self.installdir)],
+                 cwd=builddir)
```

Comment:
I suggest adding a quick cleanup step at the end of this function to make the installed binaries have useful shebangs:

```python
# Fix all shebangs to use the in-snap ruby
file_utils.replace_in_file(self.installdir, re.compile(r''),
                           re.compile(r'^#!.*ruby'),
                           r'#!/usr/bin/env ruby')
```

That allows one to use, say, the `gem` command in the `build` scriptlet. Note that you'll need to import `file_utils` from `snapcraft`.